parquet-variant/src/main/java/org/apache/parquet/variant/VariantElementConverter.java

```diff
@@ -0,0 +1,668 @@
+      // more values from the `value` field, then finalize. If the value was not an object, fields will be null.
+      objectBuilder = ((VariantObjectConverter) converters[typedValueIdx]).getObjectBuilder();
+    }
+
+    // If objectBuilder is non-null, then we have a partially complete object ready to write.
+    // Otherwise, the typed_value converter should have written something if it was non-null.
+    boolean hasTypedValue = objectBuilder != null || (startWritePos != builder.getWritePos());
+
+    if (valueIdx >= 0) {
+      variantValue = ((VariantValueConverter) converters[valueIdx]).getValue();
+    }
+    if (variantValue != null) {
+      if (!hasTypedValue) {
+        // Nothing else was added. We can directly append this value.
+        builder.shallowAppendVariant(variantValue.toByteBuffer());
+      } else {
+        // Both value and typed_value were non-null. This is only valid for an object.
+        Variant value =
+            new Variant(variantValue.toByteBuffer(), this.holder.topLevelHolder.metadata.toByteBuffer());
+        Variant.Type basicType = value.getType();
+        if (hasTypedValue && basicType != Variant.Type.OBJECT) {
+          throw new IllegalArgumentException("Invalid variant, conflicting value and typed_value");
+        }
+
+        for (int i = 0; i < value.numObjectElements(); i++) {
+          Variant.ObjectField field = value.getFieldAtIndex(i);
+          if (shreddedObjectKeys.contains(field.key)) {
+            // Skip any field ID that is also in the typed schema. This check is needed because readers with
+            // pushdown may not look at the value column, causing inconsistent results if a writer puth a
+            // given key
+            // only in the value column when it was present in the typed_value schema.
+            // Alternatively, we could fail at this point, since the shredding is invalid according to the
+            // spec.
+            continue;
+          }
+          objectBuilder.appendKey(field.key);
+          objectBuilder.shallowAppendVariant(field.value.getValueRawBytes());
+        }
+        builder.endObject();
+      }
+    } else if (objectBuilder != null) {
+      // typed_value was an object, and there's nothing left to append.
+      builder.endObject();
+    } else if (!hasTypedValue) {
+      // There was no value or typed_value.
+      if (objectFieldName != null) {
+        // Missing field.
+        ((VariantObjectBuilder) this.holder.builder).dropLastKey();
+      } else {
+        // For arrays and top-level fields, the spec considers this invalid, but recommends using VariantNull.
+        builder.appendNull();
+      }
+    }
+  }
+
+  /**
+   * Converter for the metadata column. It sets the current metadata in the parent converter,
+   * so that it can be used by the typed_value converter on the same row.
+   */
+  static class VariantMetadataConverter extends PrimitiveConverter implements VariantConverter {
```

Comment:
Why is this an inner class of `VariantElementConverter` when it doesn't extend `VariantElementConverter`? This is independent and seems like it should be completely separate.

Also, why not create/extend a `BinaryConverter` interface to avoid duplicating the dictionary code? That would make sense to me.